Implement textDocument/rename

tgodzik opened the pull request:
Previously, rename was not supported in any way. User would have to manually rename everything by hand and that could prove really tedious. Now, we offer renaming of any symbol with a small number of exclusions:
- unapply
- equals
- hashCode
- unary_!
- !

We can also add/remove other exclusions if needed.

There are also some special cases and several approaches that need to be mentioned:

1. When apply is renamed we add full name to each invocation

![renamed](https://user-images.githubusercontent.com/3807253/68298589-0eb61900-009a-11ea-8e61-e5c00141896b.gif)

2. Symbols can't be renamed while compiling, since we need full and correct semanticDB information - an error will pop up as a message.
3. Methods ending with colon can only be renamed to another name ending with colon. 
4. Renaming an overriden method will rename both parent symbol and current symbol:
![rename-inherit](https://user-images.githubusercontent.com/3807253/68299380-df081080-009b-11ea-8da9-177ece31d530.gif)

5. Symbols from outside workspace cannot be renamed.
6. Companion objects and classes are renamed.
![companion](https://user-images.githubusercontent.com/3807253/68310441-4fb92800-00b0-11ea-9357-7e4529743973.gif)

Currently not working:
- type reference such as `classOf[MyClass]` or `[T <: MyClass]`
- we don't check operator precedence when they are renamed
- we rename all parent's overriden methods by default - we might want to ask about that, however MessageRequest doesn't seem a pretty way to do it (might be mistaken).

Fixes https://github.com/scalameta/metals/issues/679

olafurpg commented on metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala:
```diff
@@ -0,0 +1,275 @@
+    }
+    val colonNotAllowed = name.endsWith(":") && !newName.endsWith(":")
+    if (colonNotAllowed) {
+      forbiddenColonRename(name, newName)
```
Calling this message has no side-effect
```suggestion
      client.logMessage(forbiddenColonRename(name, newName))
```

tgodzik replied:
Fixed

olafurpg commented on metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala:
```diff
@@ -0,0 +1,275 @@
+
+  private def symbolIsLocal(symbol: String): Boolean = {
+
+    lazy val isFromWorkspace = index
```
Instead of lazy val you can write a method
```suggestion
    def isFromWorkspace = index
```

tgodzik replied:
Fixed

olafurpg commented on tests/unit/src/test/scala/tests/RenameSuite.scala:
```diff
@@ -0,0 +1,432 @@
+       |
+       |object Main {
+       |  val a = new Alphabet {
+       |    override def <<me@@thod>>(adf: String): Int = 321
```
😍

olafurpg commented on tests/unit/src/test/scala/tests/RenameSuite.scala:
```diff
@@ -0,0 +1,432 @@
+             |}
+             |${input.replaceAll(allMarkersRegex, "")}""".stripMargin
+        )
+        _ <- Future.sequence(
```
To stress that we correctly use edit distance, can we artificially add a newline to the top of all open files before applying the rename? The output of the rewrites should be the same, but it will stress that we use token edit distance since the line numbers have shifted.

tgodzik replied:
Each file is now modified in tests.

olafurpg commented on tests/unit/src/test/scala/tests/RenameSuite.scala:
```diff
@@ -0,0 +1,432 @@
+      for {
+        _ <- server.initialize(
+          s"""/metals.json
+             |{"a":
```
Can we add test cases that stress the following scenarios?

* Rename symbol across different build targets
* Rename symbol across Scala/Java (just document that Java files are left unchanged)
* Rename symbol in project with compile error still works (we use an old semanticdb snapshot)

tgodzik replied:
Sure, will add that to the tests. It would be super useful to have that explicit.

tgodzik replied:
> Rename symbol across different build targets

Added 

> Rename symbol across Scala/Java

Also added

> Rename symbol in project with compile error still works

Added with a type error.

tgodzik commented:
Thanks @olafurpg for the review! I will try to address the comments soon, though I am thinking we should merge this after the next release so that we introduce implementation and test/run first. What do you think?

> Did you look into implementing textDocument/prepareRename? https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#textDocument_prepareRename I believe it could help provide faster validation to the user but I'm not 100% sure how it works.

I am not  sure if it helps us a lot. I was hoping there would be some kind of additional window or something like that - it would enable use to choose exact options for rename. However, currently it will do a very similar thing to what we do in rename. I will see if it helps us now.

> Can we use the `documentChanges` field in `WorkspaceEdit` to rename the files? One feature I love in IntelliJ is that it renames the file to match the classname when that class/companion is the only toplevel definition in the file.

Och, I wanted to do that, but totally forgot. Should be simple enough to do.

gabro commented:
> I am not sure if it helps us a lot. I was hoping there would be some kind of additional window or something like that - it would enable use to choose exact options for rename. However, currently it will do a very similar thing to what we do in rename. I will see if it helps us now.
> 
> >

As far I as understand, `prepareRename` may enable us to reject a rename before the rename input is prompted to the user. With the current flow, the user will first input the new name, and then get an error, which can be slightly annoying.

tgodzik commented:
> > I am not sure if it helps us a lot. I was hoping there would be some kind of additional window or something like that - it would enable use to choose exact options for rename. However, currently it will do a very similar thing to what we do in rename. I will see if it helps us now.
> > >
> 
> As far I as understand, `prepareRename` may enable us to reject a rename before the rename input is prompted to the user. With the current flow, the user will first input the new name, and then get an error, which can be slightly annoying.

I will check that! That would be much more useful indeed, haven't thought of that.

olafurpg commented:
Yes, the idea with `prepareRename` would be to give the user faster feedback on when the rename is not supported, ideally, before they type out the new name. For example, the validation to ensure we don't rename during compilation, that we don't rename `equals`, etc.

olafurpg commented:
>  I am thinking we should merge this after the next release so that we introduce implementation and test/run first. What do you think?

I'm not sure if that's necessary. From the looks of it, this PR is quite complete already with a solid implementation and test cases. I don't think we have to rush for the next release, it's OK if we wait a few more weeks before having a large stable release with a lot of the exciting stuff that's is almost ready (rename symbol, worksheets, breakpoint debugging, pants support)

mdedetrich commented:
> I'm not sure if that's necessary. From the looks of it, this PR is quite complete already with a solid implementation and test cases. I don't think we have to rush for the next release, it's OK if we wait a few more weeks before having a large stable release with a lot of the exciting stuff that's is almost ready (rename symbol, worksheets, breakpoint debugging, pants support)

I am getting so excited right now...

tgodzik commented:
@olafurpg I added some more tests as suggested and file renaming in case that a top level symbol matches the file name.

Let me know if there is something more to fix!

olafurpg commented on metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala:
```diff
@@ -0,0 +1,342 @@
+      index
+        .definition(MSymbol(symbol))
+        .exists { definition =>
+          workspace.toNIO.getFileSystem == definition.path.toNIO.getFileSystem &&
```
Duplicate logic from `isDependencySource` in https://github.com/scalameta/metals/blob/f54ca4951c4f0ba16b87c13d7308f475305ac81f/metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala#L252

tgodzik replied:
The second part is negated, so I didn't really see any sensible way to reuse it. I will move ti the extension method and extract the two conditions.

olafurpg commented on tests/unit/src/main/scala/tests/TestRanges.scala:
```diff
@@ -27,4 +30,23 @@ object TestRanges extends RangeReplace {
+        .asScala
+        .find(
+          change =>
+            change.isLeft && change.getLeft.getTextDocument.getUri
```
nit: 
```suggestion
            change.isLeft &&
            change.getLeft.getTextDocument.getUri.contains(file)
```

Scalafmt doesn't insert newlines in infix operators

tgodzik replied:
Fixed

olafurpg commented on tests/unit/src/test/scala/tests/RenameSuite.scala:
```diff
@@ -0,0 +1,519 @@
+       |package a
+       |trait <<A@@BC>>
+       |trait Alphabet{
+       |  this: <<ABC>> =>
```
:+1:

olafurpg commented on tests/unit/src/test/scala/tests/RenameSuite.scala:
```diff
@@ -0,0 +1,519 @@
+       |/a/src/main/scala/a/Main.scala
+       |package a
+       |object Main{
+       |  val other = new <<Oth@@er>>()
```
Should we consider erroring in this case?

tgodzik replied:
We could, but the other way would be to use Metals to at least rename Scala occurrences and rename the rest by hand or via different LSP. 

I thinking maybe add a warning in that case?

tgodzik replied:
I added a warning in a case like that.

olafurpg commented on tests/unit/src/test/scala/tests/RenameSuite.scala:
```diff
@@ -0,0 +1,519 @@
+       |""".stripMargin,
+    newName = "Bar"
+  )
+
```
Missing test case: named parameter for case class fields

```scala
case class Name(value: String)
Name(value = "42")
  .copy(value = "43")
  .value
new Name(value = "44")
```

tgodzik replied:
It seems that references are not being found for 42 nad 44. Added that test case.

tgodzik replied:
Seems to be the same issue as https://github.com/scalameta/scalameta/issues/1169

tgodzik commented:
There are 3 test cases that are not behaving correctly due to some issues in SemanticDB. I will create the issues there and add information about them to this PR.

@sswistun-vl might be able to work on some of those issues.

The `macro-annotation` test case might be probably fixed by a work around, which I will try to follow up with separately.

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -250,10 +250,20 @@ object MetalsEnrichments
+      isLocalFileSystem(workspace) &&
+        isInReadonlyDirectory(workspace)
+
+    def isWorkspaceSource(workspace: AbsolutePath): Boolean =
```
`path.startsWith(workspace)` should strictly speaking be a condition as well

tgodzik replied:
Added!

tgodzik commented:
@olafurpg Thanks again for the review! I fixed the issue with scalafix and added a warning when renaming a symbol from a java file. 

I agree that we can merge this PR and rework the last issues later on. For sure some things will come up when people start using the feature.

olafurpg commented:
I'm not sure why but the worksheet completion test seems flaky on Windows

```
X tests.worksheets.WorksheetLspSuite.completion 5755ms 
```
Might be best to disable it for now and I'll open up a ticket that I will followup on later.

olafurpg commented:
I can disable it in a followup PR, feel free to merge despite CI failure

tgodzik commented:
> I can disable it in a followup PR, feel free to merge despite CI failure

I can try to take a look at some point. I have everything setup on a Windows machine at home.

ckipp01 commented:
Really excited to start using this! 🎉
